In [1]:
%load_ext watermark
%watermark -a "Manuel Elias Orellana Lavayen" -d -v -iv

Author: Manuel Elias Orellana Lavayen

Date: 2026-10-06

Python implementation: CPython
Python version       : 3.12.10
IPython version      : 9.17.1

debugpy: 1.8.22



## Importando Librerias

In [5]:
import numpy as np
import pandas as pd
import polars as pl
from FuncionesModelos import modelos_regresion, nwrmsle
from sklearn.metrics import mean_absolute_error, r2_score, max_error, median_absolute_error, root_mean_squared_error
from xgboost import XGBRegressor

from IPython.display import display
pd.options.display.float_format = lambda x: f"{x:,.4f}" if abs(x) < 100 else f"{x:,.0f}"

from polars.lazyframe.engine_config import GPUEngine
config_gpu = GPUEngine(raise_on_fail=True)

from pathlib import Path
ruta_base = Path.cwd()
ruta_base

PosixPath('/espacio_trabajo/Analisis y entrenamiento de modelos')

## Cargando Dataset

In [6]:
variables_categoricas = ["numero_tienda", "cluster", "clase_producto", "familia_producto", "ciudad", "provincia_estado", "tipo_tienda", "numero_articulo"]

otras_variables = ["mes_seno", "mes_coseno", "dia_semana_seno", "dia_semana_coseno", "semana_seno", "semana_coseno", "en_promocion", "perecedero", "es_feriado_nacional", "es_feriado_local", "es_feriado_regional", "es_evento", "unidades_vendidas_ayer", "unidades_vendidas_promedio_semanal", "es_quincena", "petroleo_promedio_3meses"]

variable_predictora = ["unidades_vendidas"]

variables_df = variables_categoricas + otras_variables + variable_predictora

df = (pl.scan_parquet(ruta_base / "Datasets/Concatenados/Limpios/train3.parquet")
      .select(variables_df + ["año"]))

### Modelos

In [4]:
diccionario_modelos_experimentales_regresion = modelos_regresion(lista_clusters=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17], df= df, periodos_entrenamiento = [2013, 2014, 2015, 2016], periodo_testeto= [2017], pesos_perecedero = False)

------------------------cluster 1------------------------
Descripción de datos


statistic,value
str,f64
"""count""",2.7839e6
"""null_count""",0.0
"""mean""",3.48556
"""std""",10.768795
"""min""",0.0
"""25%""",0.0
"""50%""",0.0
"""75%""",3.0
"""max""",2987.0


Métricas


,Resultados
Predicción Maxima,506
Predicción Minima,0.0007
Ventas Reales,"9,703,454"
Ventas Predichas,"9,641,959"
Mean Absolute Error,1.8808
Median Absolute Error,0.6892
RMSE,6.1961
Max Error,"2,921"
R2,0.6689
NWRMSLE,0.5796


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"1,409,352","901,191","373,434","67,478","13,270","4,540",33.0000
porcentaje,50.6251,32.3715,13.4141,2.4239,0.4767,0.1631,0.0012
MAE,0.6845,1.5239,4.1150,10.5764,22.9721,60.7774,653
RMSE,1.7558,2.5031,5.6330,13.6989,29.2556,85.0680,917
Ventas Reales,0.0000,"2,157,516","3,762,371","2,028,360","899,332","741,993","28,246"
Ventas Predichas,"964,692","2,599,664","3,131,492","1,627,290","705,863","505,983","6,709"
NWRMSLE,0.6128,0.5134,0.5802,0.6193,0.6523,0.9270,2.2557


------------------------cluster 2------------------------
Descripción de datos


statistic,value
str,f64
"""count""",1.8532e6
"""null_count""",0.0
"""mean""",3.37369
"""std""",10.302854
"""min""",0.0
"""25%""",0.0
"""50%""",1.0
"""75%""",4.0
"""max""",2948.0


Métricas


,Resultados
Predicción Maxima,317
Predicción Minima,0.0009
Ventas Reales,"6,252,119"
Ventas Predichas,"6,307,802"
Mean Absolute Error,1.8865
Median Absolute Error,0.8024
RMSE,6.5388
Max Error,"2,884"
R2,0.5972
NWRMSLE,0.5948


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"873,546","662,163","259,408","39,260","5,951","2,477",32.0000
porcentaje,47.1372,35.7308,13.9978,2.1185,0.3211,0.1337,0.0017
MAE,0.7601,1.5143,4.0777,10.3876,22.2911,71.7669,768
RMSE,1.7426,2.4086,5.4379,13.1663,28.4914,103,920
Ventas Reales,0.0000,"1,599,848","2,568,330","1,175,992","397,824","432,394","26,418"
Ventas Predichas,"663,985","1,942,878","2,112,150","935,291","312,099","279,823","1,843"
NWRMSLE,0.6516,0.5072,0.5841,0.6250,0.6748,1.1281,2.9634


------------------------cluster 3------------------------
Descripción de datos


statistic,value
str,f64
"""count""",6.4862e6
"""null_count""",0.0
"""mean""",2.003149
"""std""",22.457169
"""min""",0.0
"""25%""",0.0
"""50%""",0.0
"""75%""",1.0
"""max""",17146.0


Métricas


,Resultados
Predicción Maxima,350
Predicción Minima,0.0001
Ventas Reales,"12,992,830"
Ventas Predichas,"12,449,250"
Mean Absolute Error,1.3250
Median Absolute Error,0.0536
RMSE,21.4419
Max Error,"17,061"
R2,0.0884
NWRMSLE,0.4797


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"4,388,265","1,527,494","453,946","74,137","14,937","5,130",617
porcentaje,67.6554,23.5499,6.9986,1.1430,0.2303,0.0791,0.0095
MAE,0.3696,1.4483,4.4416,12.2314,28.1243,104,"1,362"
RMSE,1.3252,2.5444,6.3277,16.1613,35.7692,139,"2,134"
Ventas Reales,0.0000,"3,450,941","4,454,185","2,242,673","1,011,416","886,596","877,357"
Ventas Predichas,"1,621,733","4,055,739","3,716,329","1,801,524","737,702","396,947","37,151"
NWRMSLE,0.4383,0.5050,0.6356,0.7229,0.8633,1.4622,3.4577


------------------------cluster 4------------------------
Descripción de datos


statistic,value
str,f64
"""count""",2.7798e6
"""null_count""",0.0
"""mean""",2.964018
"""std""",16.579468
"""min""",0.0
"""25%""",0.0
"""50%""",0.0
"""75%""",3.0
"""max""",8985.0


Métricas


,Resultados
Predicción Maxima,236
Predicción Minima,0.0005
Ventas Reales,"8,239,380"
Ventas Predichas,"8,077,334"
Mean Absolute Error,1.7552
Median Absolute Error,0.7000
RMSE,15.2261
Max Error,"8,970"
R2,0.1566
NWRMSLE,0.5742


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"1,393,101","975,605","336,491","49,119","8,029","1,926",178
porcentaje,50.1152,35.0962,12.1049,1.7670,0.2888,0.0693,0.0064
MAE,0.6784,1.4364,4.0602,10.4803,23.0679,105,"1,276"
RMSE,1.5767,2.2765,5.4111,13.2302,28.9615,141,"1,793"
Ventas Reales,0.0000,"2,320,920","3,292,708","1,468,679","535,269","318,798","235,676"
Ventas Predichas,"945,062","2,725,434","2,654,057","1,153,871","391,045","121,196","8,522"
NWRMSLE,0.6110,0.4936,0.5996,0.6319,0.7135,1.6903,3.4855


------------------------cluster 5------------------------
Descripción de datos


statistic,value
str,f64
"""count""",926600.0
"""null_count""",0.0
"""mean""",11.321298
"""std""",31.70281
"""min""",0.0
"""25%""",0.0
"""50%""",4.0
"""75%""",12.0
"""max""",3556.0


Métricas


,Resultados
Predicción Maxima,"1,103"
Predicción Minima,0.0013
Ventas Reales,"10,490,319"
Ventas Predichas,"10,400,599"
Mean Absolute Error,4.5272
Median Absolute Error,1.6797
RMSE,17.0399
Max Error,"3,316"
R2,0.7111
NWRMSLE,0.6193


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"291,724","246,847","258,563","89,612","24,271","11,765",580
porcentaje,31.4833,26.6401,27.9045,9.6711,2.6194,1.2697,0.0626
MAE,1.0581,2.4441,4.2770,9.6038,20.1483,52.8031,317
RMSE,3.1914,4.8145,6.4035,13.5966,27.6457,80.8464,486
Ventas Reales,0.0000,"668,528","2,874,684","2,768,094","1,669,736","2,035,800","432,920"
Ventas Predichas,"308,681","1,112,566","2,997,225","2,492,317","1,463,890","1,702,721","271,522"
NWRMSLE,0.7607,0.6299,0.4684,0.4809,0.5130,0.5516,1.1286


------------------------cluster 6------------------------
Descripción de datos


statistic,value
str,f64
"""count""",5.5596e6
"""null_count""",0.0
"""mean""",3.940232
"""std""",18.677309
"""min""",0.0
"""25%""",0.0
"""50%""",0.0
"""75%""",4.0
"""max""",14483.0


Métricas


,Resultados
Predicción Maxima,654
Predicción Minima,0.0002
Ventas Reales,"21,906,114"
Ventas Predichas,"21,825,074"
Mean Absolute Error,2.1643
Median Absolute Error,0.6437
RMSE,15.2238
Max Error,"14,469"
R2,0.3356
NWRMSLE,0.5650


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"2,820,300","1,711,276","802,266","157,368","32,618","11,265",493
porcentaje,50.7285,30.7806,14.4303,2.8306,0.5867,0.2026,0.0089
MAE,0.6159,1.7298,4.2713,11.2042,24.9924,73.7025,892
RMSE,2.0500,2.9941,6.1920,15.0321,32.1845,99.8250,"1,458"
Ventas Reales,0.0000,"4,230,023","8,143,152","4,767,566","2,215,022","1,918,616","529,114"
Ventas Predichas,"1,736,939","5,701,267","7,252,542","3,971,559","1,721,969","1,227,943","89,660"
NWRMSLE,0.5729,0.5355,0.5623,0.6302,0.7073,1.0285,2.4061


------------------------cluster 7------------------------
Descripción de datos


statistic,value
str,f64
"""count""",1.8532e6
"""null_count""",0.0
"""mean""",1.826377
"""std""",11.022691
"""min""",0.0
"""25%""",0.0
"""50%""",0.0
"""75%""",2.0
"""max""",4561.0


Métricas


,Resultados
Predicción Maxima,377
Predicción Minima,0.0007
Ventas Reales,"3,384,640"
Ventas Predichas,"3,332,844"
Mean Absolute Error,1.2491
Median Absolute Error,0.0966
RMSE,9.3009
Max Error,"4,295"
R2,0.2880
NWRMSLE,0.5062


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"1,195,948","493,838","135,192","16,644","2,836",980,63.0000
porcentaje,64.5342,26.6479,7.2951,0.8981,0.1530,0.0529,0.0034
MAE,0.4256,1.4030,4.3669,14.0002,33.8605,117,"1,076"
RMSE,1.4125,2.2627,5.9042,17.9583,39.9427,147,"1,378"
Ventas Reales,0.0000,"1,140,159","1,291,747","492,872","192,231","175,778","76,904"
Ventas Predichas,"509,045","1,330,581","961,680","327,029","112,094","65,401","9,141"
NWRMSLE,0.4778,0.4890,0.6711,0.9061,1.1271,1.7047,2.8824


------------------------cluster 8------------------------
Descripción de datos


statistic,value
str,f64
"""count""",2.7798e6
"""null_count""",0.0
"""mean""",6.743296
"""std""",19.695036
"""min""",0.0
"""25%""",0.0
"""50%""",2.0
"""75%""",7.0
"""max""",2367.0


Métricas


,Resultados
Predicción Maxima,661
Predicción Minima,0.0005
Ventas Reales,"18,745,012"
Ventas Predichas,"18,656,584"
Mean Absolute Error,2.8916
Median Absolute Error,1.1091
RMSE,9.7263
Max Error,"2,170"
R2,0.7561
NWRMSLE,0.5914


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"1,067,422","893,802","608,603","145,739","35,852","15,371",483
porcentaje,38.3992,32.1535,21.8938,5.2428,1.2897,0.5530,0.0174
MAE,0.8427,1.8594,3.9059,9.5490,19.1705,53.6878,286
RMSE,2.3581,3.1933,5.5610,12.7295,25.5361,80.2341,412
Ventas Reales,0.0000,"2,317,164","6,384,982","4,460,559","2,467,026","2,687,366","309,472"
Ventas Predichas,"899,553","3,251,156","5,959,602","3,894,962","2,151,016","2,184,247","174,612"
NWRMSLE,0.6831,0.5509,0.4897,0.5061,0.5043,0.6601,1.2939


------------------------cluster 9------------------------
Descripción de datos


statistic,value
str,f64
"""count""",1.8532e6
"""null_count""",0.0
"""mean""",2.715687
"""std""",8.599359
"""min""",0.0
"""25%""",0.0
"""50%""",0.0
"""75%""",3.0
"""max""",1093.0


Métricas


,Resultados
Predicción Maxima,300
Predicción Minima,0.0005
Ventas Reales,"5,032,712"
Ventas Predichas,"5,030,810"
Mean Absolute Error,1.5926
Median Absolute Error,0.6453
RMSE,5.5623
Max Error,"1,046"
R2,0.5816
NWRMSLE,0.5611


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"949,299","648,035","210,849","27,527","3,953","1,648",65.0000
porcentaje,51.2249,34.9684,11.3776,1.4854,0.2133,0.0889,0.0035
MAE,0.6504,1.3966,3.9721,10.7021,25.9518,83.7738,425
RMSE,2.1263,2.1689,5.2289,13.5113,32.1166,116,445
Ventas Reales,0.0000,"1,541,586","2,047,900","808,655","271,748","271,813","37,773"
Ventas Predichas,"617,460","1,781,700","1,618,937","606,548","195,177","138,997","10,154"
NWRMSLE,0.5945,0.4828,0.5945,0.6713,0.8471,1.2137,1.7462


------------------------cluster 10------------------------
Descripción de datos


statistic,value
str,f64
"""count""",5.5596e6
"""null_count""",0.0
"""mean""",2.95895
"""std""",13.056398
"""min""",0.0
"""25%""",0.0
"""50%""",0.0
"""75%""",3.0
"""max""",10700.0


Métricas


,Resultados
Predicción Maxima,363
Predicción Minima,0.0002
Ventas Reales,"16,450,587"
Ventas Predichas,"16,270,916"
Mean Absolute Error,1.7580
Median Absolute Error,0.4865
RMSE,10.5358
Max Error,"10,682"
R2,0.3488
NWRMSLE,0.5510


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"3,140,851","1,644,204","606,710","110,174","24,430","7,429",197
porcentaje,56.4942,29.5741,10.9128,1.9817,0.4394,0.1336,0.0035
MAE,0.5527,1.5534,4.3780,11.9557,25.4023,84.8341,989
RMSE,1.6902,2.7167,6.2176,15.5220,31.9216,112,"1,473"
Ventas Reales,0.0000,"3,856,042","6,072,924","3,347,980","1,656,518","1,203,082","211,488"
Ventas Predichas,"1,736,007","4,722,554","5,133,635","2,678,293","1,236,164","623,796","16,697"
NWRMSLE,0.5425,0.5193,0.6111,0.6971,0.7595,1.1698,3.0723


------------------------cluster 11------------------------
Descripción de datos


statistic,value
str,f64
"""count""",2.337e6
"""null_count""",0.0
"""mean""",8.87335
"""std""",23.572952
"""min""",0.0
"""25%""",0.0
"""50%""",3.0
"""75%""",9.0
"""max""",3012.0


Métricas


,Resultados
Predicción Maxima,681
Predicción Minima,0.0012
Ventas Reales,"20,737,014"
Ventas Predichas,"20,457,608"
Mean Absolute Error,3.6806
Median Absolute Error,1.4086
RMSE,12.6691
Max Error,"2,779"
R2,0.7112
NWRMSLE,0.6094


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"801,222","674,765","608,277","177,382","47,009","18,242",630
porcentaje,34.2842,28.8731,26.0281,7.5902,2.0115,0.7806,0.0270
MAE,0.9397,2.2015,4.0606,9.7368,19.0339,54.4741,356
RMSE,2.7807,3.7925,6.0303,13.0554,24.6845,82.2954,493
Ventas Reales,0.0000,"1,810,221","6,604,272","5,446,201","3,225,730","3,125,043","424,073"
Ventas Predichas,"752,896","2,810,725","6,525,808","4,804,100","2,786,092","2,449,601","202,173"
NWRMSLE,0.7182,0.6014,0.4813,0.5096,0.5059,0.6457,1.4221


------------------------cluster 12------------------------
Descripción de datos


statistic,value
str,f64
"""count""",926600.0
"""null_count""",0.0
"""mean""",3.634039
"""std""",13.068291
"""min""",0.0
"""25%""",0.0
"""50%""",0.0
"""75%""",4.0
"""max""",3214.0


Métricas


,Resultados
Predicción Maxima,419
Predicción Minima,0.0009
Ventas Reales,"3,367,300"
Ventas Predichas,"3,378,233"
Mean Absolute Error,1.9224
Median Absolute Error,0.3319
RMSE,9.6869
Max Error,"3,153"
R2,0.4505
NWRMSLE,0.5265


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"513,248","242,703","136,558","24,852","4,756","1,312",51.0000
porcentaje,55.3905,26.1929,14.7375,2.6821,0.5133,0.1416,0.0055
MAE,0.4574,1.8841,4.0569,10.9264,23.4093,75.6644,851
RMSE,1.7070,3.1112,5.8314,14.0357,29.7444,103,"1,041"
Ventas Reales,0.0000,"633,875","1,378,292","749,907","318,406","223,650","47,083"
Ventas Predichas,"234,782","908,082","1,217,243","611,537","231,107","151,732","3,696"
NWRMSLE,0.5013,0.5542,0.5306,0.6098,0.6611,1.0550,3.1107


------------------------cluster 13------------------------
Descripción de datos


statistic,value
str,f64
"""count""",3.7064e6
"""null_count""",0.0
"""mean""",3.495446
"""std""",9.674891
"""min""",0.0
"""25%""",0.0
"""50%""",1.0
"""75%""",4.0
"""max""",1593.0


Métricas


,Resultados
Predicción Maxima,275
Predicción Minima,0.0003
Ventas Reales,"12,955,519"
Ventas Predichas,"12,830,484"
Mean Absolute Error,1.8715
Median Absolute Error,0.7668
RMSE,5.8130
Max Error,"1,536"
R2,0.6390
NWRMSLE,0.5744


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"1,740,767","1,303,724","536,607","84,591","14,958","4,804",74.0000
porcentaje,46.9665,35.1749,14.4778,2.2823,0.4036,0.1296,0.0020
MAE,0.6951,1.5117,3.9341,10.4244,22.7148,66.6107,585
RMSE,1.6444,2.3673,5.2454,13.2701,28.3999,94.0532,621
Ventas Reales,0.0000,"3,191,507","5,320,223","2,533,628","1,004,843","738,856","51,658"
Ventas Predichas,"1,210,002","3,895,259","4,387,235","2,004,109","748,767","443,350","8,348"
NWRMSLE,0.6201,0.5019,0.5636,0.6202,0.6935,1.0970,2.2018


------------------------cluster 14------------------------
Descripción de datos


statistic,value
str,f64
"""count""",3.7064e6
"""null_count""",0.0
"""mean""",7.349153
"""std""",21.881159
"""min""",0.0
"""25%""",0.0
"""50%""",2.0
"""75%""",8.0
"""max""",6932.0


Métricas


,Resultados
Predicción Maxima,606
Predicción Minima,0.0002
Ventas Reales,"27,238,896"
Ventas Predichas,"26,992,460"
Mean Absolute Error,3.2517
Median Absolute Error,1.2794
RMSE,14.1401
Max Error,"6,882"
R2,0.5824
NWRMSLE,0.6053


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"1,329,148","1,144,394","917,528","225,862","49,326","20,459",647
porcentaje,35.8609,30.8762,24.7552,6.0938,1.3308,0.5520,0.0175
MAE,0.9166,2.0531,3.9539,9.6654,20.4739,58.0381,619
RMSE,2.8974,3.5471,5.8684,13.1687,26.9029,84.3787,845
Ventas Reales,0.0000,"3,077,631","9,737,574","6,853,669","3,358,383","3,537,958","541,785"
Ventas Predichas,"1,218,306","4,611,278","9,317,010","5,966,600","2,876,005","2,697,186","141,718"
NWRMSLE,0.7117,0.5740,0.4817,0.5149,0.5496,0.7439,1.9417


------------------------cluster 15------------------------
Descripción de datos


statistic,value
str,f64
"""count""",4.633e6
"""null_count""",0.0
"""mean""",1.901104
"""std""",7.72121
"""min""",0.0
"""25%""",0.0
"""50%""",0.0
"""75%""",2.0
"""max""",3857.0


Métricas


,Resultados
Predicción Maxima,231
Predicción Minima,0.0001
Ventas Reales,"8,807,814"
Ventas Predichas,"9,087,523"
Mean Absolute Error,1.2172
Median Absolute Error,0.0837
RMSE,5.7182
Max Error,"3,806"
R2,0.4515
NWRMSLE,0.4900


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"2,993,788","1,208,397","352,423","51,456","8,642","2,412",66.0000
porcentaje,64.6188,26.0824,7.6068,1.1106,0.1865,0.0521,0.0014
MAE,0.3952,1.4534,4.3271,12.1884,28.0372,92.7677,856
RMSE,1.3097,2.3659,5.7961,15.1560,34.0208,121,"1,032"
Ventas Reales,0.0000,"2,767,167","3,437,787","1,531,326","577,086","394,032","62,110"
Ventas Predichas,"1,183,255","3,353,214","2,813,691","1,131,143","378,650","176,342","5,591"
NWRMSLE,0.4584,0.4988,0.6240,0.7491,0.9103,1.4675,2.9174


------------------------cluster 16------------------------
Descripción de datos


statistic,value
str,f64
"""count""",926600.0
"""null_count""",0.0
"""mean""",2.778246
"""std""",8.742009
"""min""",0.0
"""25%""",0.0
"""50%""",0.0
"""75%""",3.0
"""max""",2886.0


Métricas


,Resultados
Predicción Maxima,187
Predicción Minima,0.0008
Ventas Reales,"2,574,322"
Ventas Predichas,"2,632,842"
Mean Absolute Error,1.6587
Median Absolute Error,0.6150
RMSE,6.1613
Max Error,"2,883"
R2,0.5033
NWRMSLE,0.5617


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"482,996","316,730","102,879","16,289","2,860",766,8.0000
porcentaje,52.1256,34.1820,11.1028,1.7579,0.3087,0.0827,0.0009
MAE,0.6118,1.4961,4.2579,11.0054,25.1560,75.9999,"1,170"
RMSE,1.4952,2.4423,5.6897,13.8298,30.8683,101,"1,374"
Ventas Reales,0.0000,"740,629","1,012,716","483,777","194,203","114,253","9,521"
Ventas Predichas,"295,474","917,003","839,502","367,322","133,981","56,664",165
NWRMSLE,0.5795,0.5029,0.6009,0.6770,0.8150,1.2733,4.4862


------------------------cluster 17------------------------
Descripción de datos


statistic,value
str,f64
"""count""",926600.0
"""null_count""",0.0
"""mean""",5.789366
"""std""",23.773111
"""min""",0.0
"""25%""",0.0
"""50%""",1.0
"""75%""",5.0
"""max""",4958.0


Métricas


,Resultados
Predicción Maxima,603
Predicción Minima,0.0014
Ventas Reales,"5,364,428"
Ventas Predichas,"5,477,290"
Mean Absolute Error,2.8114
Median Absolute Error,0.9979
RMSE,16.6286
Max Error,"4,869"
R2,0.5107
NWRMSLE,0.6345


Rendimiento por bloques de unidades vendidas


bloque,0,1-5,6-20,21-50,51-100,101-500,>500
cantidad,"400,095","295,777","172,424","40,278","9,807","4,484",88.0000
porcentaje,43.1788,31.9207,18.6082,4.3469,1.0584,0.4839,0.0095
MAE,1.0131,1.8324,4.2098,10.2130,21.3414,47.2011,914
RMSE,5.7111,3.3480,6.2021,14.1175,29.4206,66.2945,"1,495"
Ventas Reales,0.0000,"744,765","1,789,400","1,228,057","672,587","780,285","116,148"
Ventas Predichas,"405,354","1,022,534","1,631,416","1,063,656","591,469","685,543","35,692"
NWRMSLE,0.7243,0.5602,0.5447,0.5588,0.5366,0.5981,1.5654


## Guardando Modelos

In [12]:
for cluster, modelo in diccionario_modelos_experimentales_regresion.items():
    booster_nativo = modelo.get_booster()
    booster_nativo.save_model(f"{ruta_base}/Modelos Finales/cluster_{cluster}.json")

## Modelo General

In [3]:
def modelo_fusionado(datos: pl.DataFrame):
      lista_modelos = []
      for cluster in range(1, 18):
          modelo= XGBRegressor()
          modelo.load_model(fname= f"{ruta_base}/Modelos Finales/cluster_{cluster}.json")
          lista_modelos.append(modelo)

      # Extraemos la columna de clústeres y variables predictoras
      clusters = datos.select(["cluster"]).to_numpy()
      caracteristicas = datos.drop(["cluster"])

      # Creamos un array vacio de predicciones para ir rellenando conforme se hagan las predicciones (Asi manetenmos el orden)
      predicciones = np.full(shape= datos.shape[0], fill_value= np.nan)

      # Realizamos las predicciones con base en cada cluster
      for num_modelo, modelo in enumerate(lista_modelos):
          # Creamos una máscara para rellenar con ella el array de nulos y para seleccionar los datos que necesitamos
          mascara_cluster = (clusters == (num_modelo + 1)) # Sumamos 1 porque enumerate los enumera desde 0
          mascara_cluster = mascara_cluster.flatten() # Aplanamos la máscara bidimensional

          # Sí hay filas con respecto a este cluster, predecimos únicamente estas filas
          if mascara_cluster.any():
              predicciones[mascara_cluster] = modelo.predict(caracteristicas.filter(mascara_cluster).to_pandas())

      return predicciones

In [7]:
datos = df.filter(pl.col("año") == 2017).drop(["unidades_vendidas", "año"]).collect(engine= "streaming")
predicciones = modelo_fusionado(datos = datos)

## Métricas Modelo General

In [8]:
#Valores Reales
valores_reales = df.select(["año", "unidades_vendidas", "perecedero"]).filter(pl.col("año") == 2017).drop(["año"]).collect(engine="streaming").to_pandas()

# Para ver metricas generales del modelo general
error_absoluto_medio            = mean_absolute_error(y_pred=predicciones, y_true=valores_reales["unidades_vendidas"])
raiz_error_cuadratico_medio     = root_mean_squared_error(y_pred=predicciones, y_true=valores_reales["unidades_vendidas"])
r2                              = r2_score(y_pred=predicciones, y_true=valores_reales["unidades_vendidas"])
error_maximo                    = max_error(y_pred=predicciones, y_true=valores_reales["unidades_vendidas"])
mediana_error_absoluto          = median_absolute_error(y_pred=predicciones, y_true=valores_reales["unidades_vendidas"])
nwrmsle_metrica                 = nwrmsle(y_true=valores_reales["unidades_vendidas"],y_pred=predicciones,perecedero=valores_reales["perecedero"])

diccionario_metricas = {
    "Mean Absolute Error": error_absoluto_medio,
    "RMSE" : raiz_error_cuadratico_medio,
    "Median Absolute Error" : mediana_error_absoluto,
    "Max Error" : error_maximo,
    "R2": r2,
    "NWRMSLE": nwrmsle_metrica
}

# Para ver metricas por bloques de unidades Vendidas del modelo general
bloques = [
             ("0", 0, 0),
            ("1-5", 1, 5),
            ("6-20", 6, 20),
            ("21-50", 21, 50),
            ("51-100", 51, 100),
            ("101-500", 101, 500),
            (">500", 501, np.inf)
        ]
y_real = valores_reales["unidades_vendidas"]
y_real_perecedero = valores_reales["perecedero"]
resultados_bloques = []
for nombre_bloque, minimo, maximo in bloques:
    mascara     = ((y_real >= minimo) &(y_real <= maximo))
    y_bloque    = y_real[mascara]
    perecedero_bloque = y_real_perecedero[mascara]
    pred_bloque = predicciones[mascara]

    if len(y_bloque) == 0:
        continue

    resultados_bloques.append({
                "bloque"          : nombre_bloque,
                 "Ventas Reales"   : y_bloque.sum(),
                "Ventas Predichas": pred_bloque.sum(),
                "MAE"             : mean_absolute_error(y_bloque, pred_bloque),
                "NWRMSLE"         : nwrmsle(y_true=y_bloque, y_pred=pred_bloque, perecedero=perecedero_bloque.to_numpy())
            })

In [9]:
print("Distribución de datos")
valores_reales["unidades_vendidas"].describe()

Distribución de datos


count   49,597,700
mean        3.9164
std        17.1007
min         0.0000
25%         0.0000
50%         0.0000
75%         3.3670
max         17,146
Name: unidades_vendidas, dtype: float64

In [10]:
display(pd.DataFrame(data = diccionario_metricas, index =["Resultado"]).T)

,Resultado
Mean Absolute Error,2.1435
RMSE,13.8289
Median Absolute Error,0.6165
Max Error,"17,081"
R2,0.3460
NWRMSLE,0.5569


### Conclusiones Rendimiento General del Modelo

El rendimiento del modelo es aceptable en algunos puntos, pero flaquea en otros. Predecir las unidades vendidas de un producto en una cadena de supermercado resulta complicado por la cantidad de factores externos que influyen en el comportamiento de compras de los consumidores.

- Se presenta un MAE de 2.14, lo que considerando la distribucion de los datos parece malo, pero esta es una métrica que se ve muy afectada por valores atos debido a su naturaleza de media, lo que implica para su interpretacion correcta apoyarse de otra metrica basada en la mediana que es la mediana del error absoluto, donde podemos observar que aproximadamente más de la mitad de los datos recibieron un error menor a 1 (0.66), lo cual si tiene más concordancia con la distribucion de los datos. Esto indica que el modelo tiende a estar muy cerca de su valor verdadero.

- Se presenta un RMSE de 13.82, lo que indica que el modelo puede tener una MAE un poco confiable, pero llega a cometer errores muy grandes. Analizando la distribucion de los datos para cada cluster independiente, podemos observar que todos los modelos tuvieron problemas para que su prediccion maxima se acerque a las unidades vendidas más altas de su respectivo cluster. Esto se debe a que el modelo aprende mejor a predecir valores bajos por el exceso de ceros.

- Se presenta un R2 de 0.34, lo que indica que una poca variabilidad en las predicciones por las variables predictoras. Aunque esta metrica no está directamente relacionada con el rendimiento del modelo.

- Se presenta un NWRMSLE de 0.55, lo que púede ser aceptable. Esto nos quiere decir que nuestro modelo se equivoca un 55% en la prediccion de unidades vendidas, cuidando más el inventario de los productos perecibles

In [11]:
display(pd.DataFrame(data = resultados_bloques).set_index("bloque").T)

bloque,0,1-5,6-20,21-50,51-100,101-500,>500
Ventas Reales,0.0000,"36,248,524","70,173,240","42,387,992","20,667,360","19,586,312","4,017,748"
Ventas Predichas,"14,649,361","43,412,332","55,336,772","29,693,576","13,001,690","8,618,207","411,452"
MAE,0.5680,1.5361,4.2015,11.6265,27.5789,96.1389,838
NWRMSLE,0.5504,0.5118,0.6046,0.6894,0.7810,1.1476,2.6414


### Conclusiones Rendimiento General del Modelo Analizando por Bloques de Unidades Vendidas

#### 0 Unidades Vendidas
- El modelo resulta tener un rendimiento aceptable en su MAE y NWRMSLE cuando se trata de predecir productos con cero unidades vendidas. Aunque parece predecir muchas unidades vendidas con respecto a las reales esto principalmente sucede porque al cometer varios errores pequeños (como 0.0002), estos se van sumando y aumentan la cantidad con respecto a la real. esto puede solucionarse con un modelo de dos estapas que detecte si un producto se va a vender o no, funcionando como filtro y asi evitar esos pequeños errores. Además, esto hace que el modelo sea util par analisis, más no para producción.

#### 1-5 Unidades Vendidas
- El modelo resulta tener su mejor rendimiento cuando de predecir unidades bajas se trata, presentando un MAE de 1.5 y un NWRMSLE de 0.51, no obstante, suele predecir un poco más de lo esperado, lo que puede producir un exceso de stock, pero también hay que tener en cuenta que su NWRMSLE nos indica que gran parte de esos productos no serían perecederos

#### 6-20 Unidades Vendidas
- El modelo comienza a flaquear su rendimiento desde este punto, donde presenta un resultado aceptable en su MAE con respecto a las unidades vendidas, pero predice menos de lo esperado en las unidades vendidas, lo que podría producir substock de diversos productos

#### 21-50 y 51-100 Unidades Vendidas
- El modelo ya flaquea bastante con un NWRMSLE muy alto, pero aún presenta un rendimiento aceptable en su MAE lo que lo hace util para predecir estas unidades vendidas.

#### > 100 Unidades Vendidas
- Predecir valores elevados es la mayor debilidad del modelo. Esto surge por la gran presencia de 0 (por los productos que no se venden) y por la poca presencia de productos elevados, lo que hace que el modelo prefiera ser un poco más conservador y  no querer arriesgarse a predecir valores altos